File: jiho/level1/숫자문자열과영단어.java
//https://school.programmers.co.kr/learn/courses/30/lessons/81301

package level1;

public class 숫자문자열과영단어 {
    class Solution {
        public int solution(String s) {
            char[] arr = s.toCharArray();
            StringBuilder sb = new StringBuilder();
            for(int i = 0; i < arr.length; i++){
                if(Character.isDigit(arr[i])){
                    sb.append(arr[i]);
                } else{
                    for(int j = i; j < arr.length; j++){
                        String a = check(s.substring(i, j+1));
                        if(a != null){
                            sb.append(a);
                            i = j;
                        }
                    }
                }
            }
            int answer = Integer.parseInt(String.valueOf(sb));
            return answer;
        }

        public String check(String s){
            switch (s){
                case "zero" : return "0";
                case "one": return "1";
                case "two": return "2";
                case "three": return "3";
                case "four": return "4";
                case "five": return "5";
                case "six": return "6";
                case "seven": return "7";
                case "eight": return "8";
                case "nine": return "9";
            }
            return null;
        }
    }
}

File: ningpop/level1/숫자_문자열과_영단어.py
# 2022.08.20
# 풀이 시간: 9분 30초
# 채점 결과: 정답
# 시간복잡도: O(N)
# 문제 링크: https://school.programmers.co.kr/learn/courses/30/lessons/81301

def solution(s: str) -> int:
    word_table = { 'zero': '0', 'one': '1', 'two': '2', 'three': '3', 'four': '4',
                    'five': '5', 'six': '6', 'seven': '7', 'eight': '8', 'nine': '9' }
    
    answer = ''
    token = ''
    for c in s:
        if c.isdigit():
            answer += c
            continue
        
        token += c
        if token in word_table.keys():
            answer += word_table[token]
            token = ''

    return int(answer)

print(solution("one4seveneight")) # 1478
print(solution("23four5six7")) # 234567
print(solution("2three45sixseven")) # 234567
print(solution("123")) # 123